### ISSUE TITLE
Jittering RMPath

### ISSUE THREAD

**Issue by jan-christiansen** [Outdooractive/route-me#29]
I have drawn a simple RMPath with two points on the map. When I zoom in or out quite fast using two finger tap, the path kind of jitters and the path sometimes even moves in from the top left corner.

**Comment by trasch** [comment]
Could you create some sample app that shows the issue?

**Comment by jan-christiansen** [comment]
I have added a branch called bug that contains an example called SimplePath to my fork. If you zoom in and out using tapping you should be able to observe that the path jumps slightly.

**Comment by trasch** [comment]
Sorry for the delay, I'm a bit busy right now. I'll have a look at it next week.

**Comment by jan-christiansen** [comment]
No problem, I am very grateful for your efforts.

**Comment by trasch** [comment]
There are several problems here, I think, and I haven't tracked them down yet. (I think the root cause is some timing issue, but thats just a guess for now).

If you really need a solution, and your paths are reasonably short, you could get away with commenting out the clipping in the recalculateGeometry (the four sections with the "Clip xxx" comments), but this will break for longer paths (it definitely breaks my apps).

I think the same underlying problem occurcs in RMCircle, #38.

**Comment by jan-christiansen** [comment]
Is it possible to support you in investigating what's going on?

**Comment by jan-christiansen** [comment]
When commenting out clipping in recalculateGeometry the app receives a memory warning and crashes with the message "Low memory in the database tilecache" when zooming in. Don't know whether you had this in mind when you said the app breaks.

**Comment by trasch** [comment]
Among others, yes. I also had the issue that the path vanishes at higher zoom levels...

**Comment by jan-christiansen** [comment]
I think I have made some progress in understanding this issue. I have considered a simpler, related issue that occurs when scrolling the map at a higher zoom level. In this case the path also shortly occurs at positions it is not supposed to.

It looks like these ghost paths are caused by implicit animations of the CALayer class. Oddly these animations are neither turned off by setting the animation duration to zero nor by invoking `setDisableActions:`. Nevertheless they are turned off by setting the actions dictionary as described here: http://stackoverflow.com/questions/2244147/disabling-implicit-animations-in-calayer-setneedsdisplayinrect/2244734#2244734. If I set the `actions` property accordingly, it looks like scrolling works as expected.

When I disable all implicit animations, double tap zooming looks worse because apparently it makes use of some of the animations. As a next step I plan to check which animations are necessary for zooming and set the dictionary accordingly or adapt the dictionary when the map zooms. I would be grateful if somebody with more experience can comment on my observations before I go on.

**Comment by trasch** [comment]
From issue #46 (plancalculus):

When I tried to use explicit animations instead of implicit once to improve path drawing, I have observed a strange behavior of RMAnnotation. It looks like my annotations are set to positions far out of bounds before they are set to their designated position. For example, when zooming in by double tap, my annotation is set to position

(138693.671875, 86909.796875)

before it is set to position

(100.679443, 175.798996)

I think these position changes result in artifacts when using animations. Are these strange positions intended?

**Comment by trasch** [comment]
I think this is also the cause for the jittering paths but I haven't yet been able to find the root cause.

**Comment by jan-christiansen** [comment]
Don't know whether this is new information but one reason for strange positions seems to be that `mapScrollView.contentOffset` in `observeValueForKeyPath:ofObject:change:context:` is set wrongly. More precisely, when I add a single marker to the map, its position is set three times. The first time `mapScrollView.contentOffset` in `observeValueForKeyPath:ofObject:change:context:` is 

(140365.000000, 88086.000000) 

while the other two times it is 

(278940.000000, 175145.000000). 

As `annotation.position` in `correctScreenPosition:` is calculated by subtracting the contentOffset and all the other values are equal we get the large difference between the first position and the later ones.

**Comment by jan-christiansen** [comment]
I don't get the behavior of `observeValueForKeyPath:ofObject:change:context:`. When I perform a single double tap zoom it is called three times. I would have expected that it is called only once by the call of `zoomToRect`. Do I miss something?

**Comment by jan-christiansen** [comment]
Except for some minor shortcomings, zooming and translating paths seems to work fine in my implementation now. To overcome the problems with the observer I am relying on the `UIScrollViewDelegate` methods only. I suspect that there is a good reason to use key value observing instead, isn't there.

To solve some other problems I have made some changes to the manuel clipping of a path layer. More precisely, I use explicit animations and use the non-clipped bounds for the animation but set the bounds of the layer to the clipped bounds. I have changed the implementation in the gist https://gist.github.com/2350064 accordingly. When I find some time, I will check what happens if I use this kind of animation with the original implementation that uses the observer.

By the way, can somebody explain to me why `masksToBounds` of the superlayer does not have the same effect as clipping the bounds of the sublayer manually?

**Comment by trasch** [comment]
I'd like to see your changes and how you reimplemented the map view to only use the delegate methods, if you want. There were some reasons to use the `observeValueForKeyPath:ofObject:change:context:` method, but I honestly can't remember them any more. Maybe there was some other requirement which doesn't exist any more...

**Comment by jan-christiansen** [comment]
I have added a branch called `without_observer` to my fork of route-me. There are still some shortcomings, for example, the animation when the zoom bounces back looks ugly on the device in my case. Furthermore, I did not check whether all features are working with this implementation. I would be very happy about any kind of comment about the code.

**Comment by trasch** [comment]
I wonder if I can close this issue? I have been using it in my apps for a while and haven't had any issues so far.

**Comment by incanus** [comment]
I haven't seen any issues, but I haven't been testing extensively.

**Comment by jan-christiansen** [comment]
I have not observed any problems in my application, too. I close this issue now and we can reopen it if somebody still experiences any problems.

**Comment by lukasvd** [comment]
I just switched to this fork for evaluation. I do experience that paths vanish at higher zoom levels (around 14/15 it starts). Sometimes when I just move (scroll) the view (at the same level), they suddenly appear again (with a slight movement I can make them appear/disappear). I added the paths as RMShape (also tried the deprecated RMPath option, same results). I am testing with the 'develop' branche.
In one case I have just 2 paths (both around 9km/100points each), one of the paths disappear. In another case I have around 15 paths (also around 10km/100 points each), I experience problems with 5 or 6 of them.
